### PR TITLE
Ports SSoverlays

### PR DESCRIPTION
**Pull request by quardbreak**
DNM, opening for diff. Rude ported from Europa. Would be nice to get any help with this.

### CODE REVIEW

File: code/__defines/subsystem-priority.dm [hunk 1]
```diff
@@ -3,11 +3,12 @@
 // SS_BACKGROUND handles high server load differently than Normal and SS_TICKER do.
 // Higher priority also means a larger share of a given tick before sleep checks.
 
-#define SS_PRIORITY_DEFAULT 50          // Default priority for all processes levels
+#define SS_PRIORITY_DEFAULT 50         // Default priority for all processes levels
 
 // SS_TICKER
+#define SS_PRIORITY_OVERLAY        200 // Applies overlays. May cause overlay pop-in if it gets behind.
```
Comment (afterthought2) [inline code/__defines/subsystem-priority.dm]: I believe this is an unreasonably large number.

File: code/controllers/subsystems/overlays.dm [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+	var/init_p_len = LAZYLEN(cached_priority)  //starter pokemon
+
+	LAZYREMOVE(cached_overlays, overlays)
+	if(priority)
+		LAZYREMOVE(cached_priority, overlays)
+
+	if(NOT_QUEUED_ALREADY && ((init_o_len != LAZYLEN(cached_priority)) || (init_p_len != LAZYLEN(cached_overlays))))
+		QUEUE_FOR_COMPILE
+
+/atom/proc/add_overlay(list/overlays, priority = FALSE)
```
Comment (afterthought2) [inline code/controllers/subsystems/overlays.dm]: Confusing arg name; use `_overlays` or something slightly different.

File: code/controllers/subsystems/overlays.dm [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+//TODO: Better solution for these?
+/image/proc/add_overlay(x)
+	overlays += x
+
+/image/proc/cut_overlay(x)
+	overlays -= x
+
+/image/proc/cut_overlays(x)
+	overlays.Cut()
```
Comment (afterthought2) [inline code/controllers/subsystems/overlays.dm:218]: What's the point of these?

File: code/controllers/subsystems/overlays.dm [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+/atom
+	var/tmp/list/our_overlays	//our local copy of (non-priority) overlays without byond magic. Use procs in SSoverlays to manipulate
+	var/tmp/list/priority_overlays	//overlays that should remain on top and not normally removed when using cut_overlay functions, like c4.
```
Comment (afterthought2) [inline code/controllers/subsystems/overlays.dm:222]: But these are not tmp?

Comment (Lohikar) [reply code/controllers/subsystems/overlays.dm:222]: They're not meant to be saved anywhere or mapped in, so they're tmp.

File: code/__defines/subsystem-priority.dm (hunk 1 above)
Comment (Lohikar) [reply code/__defines/subsystem-priority.dm]: O7 sets this to 100.

File: code/controllers/subsystems/overlays.dm (hunk 2 above)
Comment (Lohikar) [reply code/controllers/subsystems/overlays.dm]: Preferably the latter, params shouldn't be underscore-prefixed.

**Review by afterthought2 — APPROVED** [review]
Seems OK to me. Would be good if someone else also reviewed.

**Review by MistakeNot4892 — APPROVED** [review]
Seems fine to me as well.

**Review by Lohikar — APPROVED** [review]

**Pull request merged**